engine/table/src/main/java/io/deephaven/engine/table/impl/select/WhereFilterAdapter.java

```diff
@@ -0,0 +1,421 @@
+    }
+
+    @Override
+    public WhereFilter visit(ColumnName columnName) {
+        return of(columnName, inverted);
+    }
+
+    @Override
+    public WhereFilter visit(FilterPattern pattern) {
+        return of(pattern, inverted);
+    }
+
+    @Override
+    public WhereFilter visit(FilterQuick quick) {
+        return inverted ? of(quick.invert(), parentDefinition) : of(quick, parentDefinition);
+    }
+
+    @Override
+    public WhereFilter visit(FilterMatches matches) {
+        return of(matches, inverted);
+    }
+
+    @Override
+    public WhereFilter visit(Function function) {
+        return of(function, inverted);
+    }
+
+    @Override
+    public WhereFilter visit(Method method) {
+        return of(method, inverted);
+    }
+
+    @Override
+    public WhereFilter visit(boolean literal) {
+        return of(literal ^ inverted);
+    }
+
+    @Override
+    public WhereFilter visit(RawString rawString) {
+        return of(rawString, inverted);
+    }
+
+    private static class FilterComparisonAdapter implements Expression.Visitor<WhereFilter> {
+
+        public static WhereFilter of(FilterComparison condition) {
+            FilterComparison preferred = condition.maybeTranspose();
+            return preferred.lhs().walk(new FilterComparisonAdapter(condition, preferred));
+        }
+
+        private final FilterComparison original;
+        private final FilterComparison preferred;
+
+        private FilterComparisonAdapter(FilterComparison original, FilterComparison preferred) {
+            this.original = Objects.requireNonNull(original);
+            this.preferred = Objects.requireNonNull(preferred);
+        }
+
+        private WhereFilter original() {
+            return WhereFilterFactory.getExpression(Strings.of(original));
+        }
+
+        @Override
+        public WhereFilter visit(ColumnName lhs) {
+            return preferred.rhs().walk(new FilterComparisonAdapter.PreferredLhsColumnRhsVisitor(lhs));
+        }
+
+        private class PreferredLhsColumnRhsVisitor
+                implements Expression.Visitor<WhereFilter>, Literal.Visitor<WhereFilter> {
+            private final ColumnName lhs;
+
+            public PreferredLhsColumnRhsVisitor(ColumnName lhs) {
+                this.lhs = Objects.requireNonNull(lhs);
+            }
+
+            @Override
+            public WhereFilter visit(ColumnName rhs) {
+                // LHS column = RHS column
+                return original();
+            }
+
+            @Override
+            public WhereFilter visit(int rhs) {
+                switch (preferred.operator()) {
+                    case EQUALS:
+                        return new MatchFilter(lhs.name(), rhs);
+                    case NOT_EQUALS:
+                        return new MatchFilter(MatchType.Inverted, lhs.name(), rhs);
+                    // Note: we can't assume IntRangeFilter - even though the rhs literal is an int, it might be against
```

Comment:
This bothers me, since this method is only reachable when you're in a DH query engine of some kind. In fact, the only way that this method can be reached at all is from a WhereFilterAdapter instance, which plainly has a TableDefinition field. I don't immediately see a downside to dropping the static modifier on this inner-inner class (just have to move/rename the "public" static `of()` of the containing inner class), or pass the reference inward, so that this knowledge is available. Since the inner type FilterComparisonAdapter is private anyway, there can be no use case where, contrary to this comment, TableDefinition isn't available, so we should always know the appropriate type?

Reply:
https://github.com/deephaven/deephaven-core/issues/3730